File: Anteater/CompassViewController.m
//
//  CompassViewController.m
//  Anteater
//
//  Created by Sam Madden on 1/29/16.
//  Copyright © 2016 Sam Madden. All rights reserved.
//

#import "CompassViewController.h"
#import "AnteaterREST.h"

#define DEGREES_TO_RADIANS(degrees) ((degrees / 180.0) * M_PI)
#define RADIANS_TO_DEGREES(radians) (radians * (180.0 / M_PI))

@interface CompassViewController ()

@end

@implementation CompassViewController {
    NSArray *_anthills;
    CLLocationManager *_mgr;
    UIImage *_image;
    BOOL gotLoc, gotHeading;
    CLLocation *_lastLoc, *_userLoc, *_targetLoc;
    CGFloat _curHeading, _lastHeading, _scale, _lastMagHeading;
    // TODO: determine use of _lastLoc, _lastHeading, _scale, and _lastMagHeading
}

- (void)viewDidLoad {
    _mgr = [[CLLocationManager alloc] init];
    _mgr.delegate = self;
    _mgr.desiredAccuracy = kCLLocationAccuracyBest;
    _mgr.distanceFilter = 0;
    _mgr.headingOrientation = CLDeviceOrientationPortrait;
    [_mgr startUpdatingHeading];
    [_mgr startUpdatingLocation];
    _anthills = @[];
    _picker.dataSource = self;
    _picker.delegate = self;
    [AnteaterREST getListOfAnthills:^(NSDictionary *hills) {
        if (hills)
            _anthills = [hills objectForKey:@"anthills"];
        [_picker reloadAllComponents];
        _targetLoc = [self curSelectedLocation];
    }];
    
    self.distanceLabel.text = @"";
    self.headingLabel.text = @"";

    [super viewDidLoad];
    // Do any additional setup after loading the view.
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (double) computeDegreesFromLocation:(CLLocation *) src toLocation:(CLLocation *) dst
{
    double lon1 = DEGREES_TO_RADIANS(src.coordinate.longitude);
    double lat1 = DEGREES_TO_RADIANS(src.coordinate.latitude);
    double lon2 = DEGREES_TO_RADIANS(dst.coordinate.longitude);
    double lat2 = DEGREES_TO_RADIANS(dst.coordinate.latitude);
    
    double angleRadians = atan2(sin(lon2-lon1)*cos(lat2), cos(lat1)*sin(lat2)-sin(lat1)*cos(lat2)*cos(lon2-lon1));
    
    return RADIANS_TO_DEGREES(angleRadians);
}

- (void) updateInterface
{
    if (gotLoc && gotHeading) {
        // both are in degrees
        double thetaOne = [self computeDegreesFromLocation:_userLoc toLocation:_targetLoc];
        double thetaTwo = _curHeading;

        double heading =  thetaOne - thetaTwo;
        double distance = [_userLoc distanceFromLocation:_targetLoc];
        
        self.distanceLabel.text = [NSString stringWithFormat: @"%4fkm", distance/1000];
        self.headingLabel.text = [NSString stringWithFormat:@"%3f%@", heading, @"\u00B0"];
        _needle.transform = CGAffineTransformMakeRotation(DEGREES_TO_RADIANS(heading));
    }
}

#pragma mark - CLLocationManagerDelegate Methods -

- (void)locationManager:(CLLocationManager *)manager didUpdateHeading:(CLHeading *)newHeading {
    _lastHeading = _curHeading;
    _curHeading = newHeading.trueHeading;
    
    gotHeading = TRUE;
   
    [self updateInterface];
}


-(void)locationManager:(CLLocationManager*)manager didUpdateLocations:(nonnull NSArray<CLLocation *> *)locations {
    _lastLoc = _userLoc;
    _userLoc = [locations lastObject];
    gotLoc = TRUE;

    [self updateInterface];
}


#pragma  mark - Picker View -

- (NSInteger) numberOfComponentsInPickerView:(UIPickerView *)pickerView {
    return 1;
}

- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
    return [_anthills count];
}

- (nullable NSString *)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component __TVOS_PROHIBITED {
    return [[_anthills objectAtIndex:row] objectForKey:@"id"];
}

-(CLLocation *) curSelectedLocation {
    NSDictionary *d = [_anthills objectAtIndex:[_picker selectedRowInComponent:0]];
    CLLocation *hill = [[CLLocation alloc] initWithLatitude:[[d objectForKey:@"lat"] floatValue] longitude:[[d objectForKey:@"lon"] floatValue]];
    return hill;

}

- (void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row inComponent:(NSInteger)component __TVOS_PROHIBITED
{
    _targetLoc = [self curSelectedLocation];
    
    [self updateInterface];
}


@end

File: Anteater/CompassViewController.swift
//
//  CompassViewController.swift
//  Anteater
//
//  Created by Ankush Gupta on 3/2/16.
//  Copyright © 2016 Sam Madden. All rights reserved.
//

import Foundation
import SwiftyJSON

@objc class CompassViewController : UIViewController, CLLocationManagerDelegate, UIPickerViewDataSource, UIPickerViewDelegate {
    
    @IBOutlet var needle : CompassView!;
    @IBOutlet var picker : UIPickerView!;
    @IBOutlet var headingLabel, distanceLabel : UILabel!
    
    var anthills: JSON = []
    var mgr: CLLocationManager?
    var image: UIImage?
    var gotLoc = false
    var gotHeading = false
    var userLoc, lastLoc, targetLoc: CLLocation?
    var curHeading = 0.0, lastHeading = 0.0
    
//    var _scale, _lastMagHeading: CGFloat?
    
    static func degToRad(x : Double) -> Double {
        return (M_PI * (x) / 180.0)
    }
    
    static func radToDeg(x :Double) -> Double{
        return ((x) * 180.0 / M_PI)
    }

	static func computeDegreesFromLocation(src: CLLocation, toLocation dst: CLLocation) -> Double {
		let lon1 = degToRad(src.coordinate.longitude)
		let lat1 = degToRad(src.coordinate.latitude)
		let lon2 = degToRad(dst.coordinate.longitude)
		let lat2 = degToRad(dst.coordinate.latitude)
        
		let angleRadians = atan2(sin(lon2-lon1)*cos(lat2), cos(lat1)*sin(lat2)-sin(lat1)*cos(lat2)*cos(lon2-lon1))
        
		return radToDeg(angleRadians)
	}
    
    override func viewDidLoad() {
        mgr = CLLocationManager()
        mgr?.delegate = self
        mgr?.desiredAccuracy = kCLLocationAccuracyBest
        mgr?.distanceFilter = 0
        mgr?.headingOrientation = CLDeviceOrientation.Portrait
        mgr?.startUpdatingHeading()
        mgr?.startUpdatingLocation()
        
        picker?.dataSource = self
        picker?.delegate = self
        
        AnteaterREST.getListOfAnthills { (data) in
            self.anthills = data["anthills"]
            self.picker?.reloadAllComponents()
            self.targetLoc = self.curSelectedLocation()
        }
        
        self.distanceLabel?.text = "Loading..."
        self.headingLabel?.text = ""
        super.viewDidLoad()
    }

    func updateInterface() {
        if (gotLoc && gotHeading) {
            if let target = targetLoc, user = userLoc{
                // both are in degrees
                let thetaOne = CompassViewController.computeDegreesFromLocation(user, toLocation: target)
                let thetaTwo = curHeading
                let heading = thetaOne - thetaTwo
                let distance = user.distanceFromLocation(target)
                
                self.distanceLabel?.text = String(format: "%4fkm", distance/1000)
                self.headingLabel?.text = String(format: "%3f°", heading)
                needle?.transform = CGAffineTransformMakeRotation(CGFloat(CompassViewController.degToRad(heading)))
            }

        }
    }
    
	func curSelectedLocation() -> CLLocation? {
		var d = anthills[picker.selectedRowInComponent(0)]
        if let lat = d["lat"].double, let lon = d["lon"].double {
            return CLLocation(latitude: lat, longitude: lon)
        } else {
            return nil
        }
	}
    
    // MARK: CLLocationManagerDelegate
    func locationManager(manager: CLLocationManager, didUpdateHeading newHeading: CLHeading) {
        lastHeading = curHeading
        curHeading = newHeading.trueHeading
        gotHeading = true
        self.updateInterface()
    }
    
    func locationManager(manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        lastLoc = userLoc
        userLoc = locations.last
        gotLoc = true
        self.updateInterface()
    }

    // MARK: UIPickerView methods
    func numberOfComponentsInPickerView(pickerView: UIPickerView) -> Int {
        return 1
    }
    
    func pickerView(pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        return anthills.count
    }

    func pickerView(pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
        if let id = anthills[row]["id"].string{
            if let description = anthills[row]["description"].string {
                return "\(id) - \(description)"
            } else {
                return id
            }
        }
        return nil
    }
    
    func pickerView(pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
        targetLoc = self.curSelectedLocation()
        self.updateInterface()
    }
}